File: src/python/xerxes/streaming/responses_api.py
# Copyright 2026 The Xerxes-Agents Author @erfanzar (Erfan Zare Chavoshi).
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     https://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
"""Translate OpenAI / OpenRouter Responses-API events into neutral chunks.

The Responses API emits a stream of typed events such as
``response.output_text.delta``, ``response.reasoning.delta``,
``response.output_item.added``, ``response.function_call_arguments.delta``,
``response.output_item.done``, and finally ``response.completed``.

:class:`ResponsesEventTranslator` consumes those raw dicts and yields
:class:`TextChunk` / :class:`ThinkingChunk` chunks while accumulating
tool-call records and usage counters on :attr:`ResponsesEventTranslator.usage`.
"""

from __future__ import annotations

from collections.abc import Iterable, Iterator
from dataclasses import dataclass, field
from typing import Any

from .events import TextChunk, ThinkingChunk


@dataclass
class ResponsesUsage:
    """Accumulated usage and tool-call summary for a Responses-API stream.

    Attributes:
        input_tokens: Prompt tokens reported by ``response.completed``.
        output_tokens: Completion tokens reported by ``response.completed``.
        cache_read_tokens: Tokens served from cache when reported.
        cache_creation_tokens: Tokens written to a new cache entry.
        tool_calls: Tool-call records assembled from ``output_item`` events.
        finish_reason: Final response status (``"stop"``, ``"error"``, …).
    """

    input_tokens: int = 0
    output_tokens: int = 0
    cache_read_tokens: int = 0
    cache_creation_tokens: int = 0
    tool_calls: list[dict[str, Any]] = field(default_factory=list)
    finish_reason: str = "stop"


class ResponsesEventTranslator:
    """Stateful translator from Responses-API events to neutral stream chunks.

    Holds an in-progress buffer of streaming tool-call argument deltas keyed
    by item id, and a :class:`ResponsesUsage` snapshot updated on completion.
    """

    def __init__(self) -> None:
        """Initialise an empty usage snapshot and tool-call argument buffer."""
        self.usage = ResponsesUsage()
        self._buffer: dict[str, dict[str, Any]] = {}

    def translate(self, events: Iterable[dict[str, Any]]) -> Iterator[TextChunk | ThinkingChunk]:
        """Walk Responses-API events and yield neutral chunks.

        Recognises ``response.output_text.delta`` and ``response.reasoning.delta``
        (yielded immediately as :class:`TextChunk` / :class:`ThinkingChunk`),
        assembles tool calls across ``output_item.added`` /
        ``function_call_arguments.delta`` / ``output_item.done`` events into
        :attr:`usage.tool_calls`, and finalises counters on
        ``response.completed`` (or marks ``finish_reason`` on errors).
        """
        for event in events:
            etype = event.get("type", "")
            if etype == "response.output_text.delta":
                delta = event.get("delta") or event.get("text", "")
                if delta:
                    yield TextChunk(delta)
            elif etype == "response.reasoning.delta":
                delta = event.get("delta") or event.get("text", "")
                if delta:
                    yield ThinkingChunk(delta)
            elif etype == "response.output_item.added":
                item = event.get("item", {})
                if item.get("type") == "tool_call":
                    self._buffer[item.get("id", "")] = {"name": item.get("name", ""), "arguments_text": ""}
            elif etype == "response.function_call_arguments.delta":
                call_id = event.get("item_id", "")
                if call_id in self._buffer:
                    self._buffer[call_id]["arguments_text"] += event.get("delta", "")
            elif etype == "response.output_item.done":
                item = event.get("item", {})
                if item.get("type") == "tool_call":
                    call_id = item.get("id", "")
                    record = self._buffer.pop(call_id, {"name": item.get("name", ""), "arguments_text": ""})
                    args_text = item.get("arguments") or record["arguments_text"]
                    self.usage.tool_calls.append(
                        {"id": call_id, "name": record["name"] or item.get("name", ""), "arguments_text": args_text}
                    )
            elif etype == "response.completed":
                resp = event.get("response", {})
                usage = resp.get("usage", {})
                self.usage.input_tokens = int(usage.get("input_tokens", 0))
                self.usage.output_tokens = int(usage.get("output_tokens", 0))
                self.usage.cache_read_tokens = int(usage.get("cache_read_tokens", 0))
                self.usage.cache_creation_tokens = int(usage.get("cache_creation_tokens", 0))
                self.usage.finish_reason = resp.get("status", "stop")
            elif etype in ("response.failed", "error"):
                self.usage.finish_reason = "error"


__all__ = ["ResponsesEventTranslator", "ResponsesUsage"]

File: src/typescript/src/streaming/responsesApi.ts
// Copyright 2026 The Xerxes-Agents Author @erfanzar (Erfan Zare Chavoshi).
// Licensed under the Apache License, Version 2.0.

import type { LlmDelta, TokenUsage } from '../llms/client.js'
import { isJsonObject, parseToolArguments, type ToolCall } from '../types/toolCalls.js'

export interface ResponsesUsage extends TokenUsage {
  readonly finishReason: string
  readonly toolCalls: readonly ToolCall[]
}

interface PendingFunctionCall {
  argumentsText: string
  readonly id: string
  name: string
}

/**
 * Stateful normalizer for streamed OpenAI Responses API events.
 *
 * It keeps partial function-call arguments out of visible model text and
 * produces the same neutral LlmDelta vocabulary consumed by the agent loop.
 */
export class ResponsesEventTranslator {
  usage: ResponsesUsage = {
    inputTokens: 0,
    outputTokens: 0,
    toolCalls: [],
    finishReason: 'stop',
  }

  private readonly pendingCalls = new Map<string, PendingFunctionCall>()

  /** Translate one decoded Responses API event into zero or more neutral deltas. */
  translate(event: Readonly<Record<string, unknown>>): LlmDelta[] {
    const type = stringValue(event.type)
    if (type === 'response.output_text.delta') {
      const text = stringValue(event.delta) || stringValue(event.text)
      return text ? [{ content: text }] : []
    }
    if (type === 'response.reasoning.delta') {
      const thinking = stringValue(event.delta) || stringValue(event.text)
      return thinking ? [{ thinking }] : []
    }
    if (type === 'response.output_item.added') {
      this.addFunctionCall(recordValue(event.item))
      return []
    }
    if (type === 'response.function_call_arguments.delta') {
      const id = stringValue(event.item_id) || stringValue(event.call_id)
      if (!id) return []
      const pending = this.pendingCalls.get(id) ?? this.createPendingCall(id)
      pending.argumentsText += stringValue(event.delta)
      this.pendingCalls.set(id, pending)
      return []
    }
    if (type === 'response.output_item.done') {
      this.completeFunctionCall(recordValue(event.item))
      return []
    }
    if (type === 'response.completed') {
      this.completeUsage(recordValue(event.response))
      return [this.completionDelta()]
    }
    if (type === 'response.failed' || type === 'error') {
      this.usage = { ...this.usage, finishReason: 'error' }
      return []
    }
    return []
  }

  /** Translate an ordered event sequence and flush any unfinished tool calls. */
  *translateAll(events: Iterable<Readonly<Record<string, unknown>>>): Generator<LlmDelta> {
    for (const event of events) {
      for (const delta of this.translate(event)) yield delta
    }
    const final = this.finish()
    if (final) yield final
  }

  /** Flush a truncated transport after it ends without response.completed. */
  finish(): LlmDelta | undefined {
    if (!this.pendingCalls.size) return undefined
    this.completePendingCalls()
    return this.completionDelta()
  }

  private addFunctionCall(item: Readonly<Record<string, unknown>>): void {
    if (!isFunctionCallItem(item)) return
    const id = stringValue(item.id) || stringValue(item.call_id)
    if (!id) return
    const pending = this.pendingCalls.get(id) ?? this.createPendingCall(id)
    const name = stringValue(item.name)
    const argumentValue = argumentsText(item.arguments)
    if (name) pending.name = name
    if (argumentValue) pending.argumentsText = argumentValue
    this.pendingCalls.set(id, pending)
  }

  private completeFunctionCall(item: Readonly<Record<string, unknown>>): void {
    if (!isFunctionCallItem(item)) return
    const id = stringValue(item.id) || stringValue(item.call_id)
    if (!id) return
    const pending = this.pendingCalls.get(id) ?? this.createPendingCall(id)
    const name = stringValue(item.name)
    const argumentValue = argumentsText(item.arguments)
    if (name) pending.name = name
    if (argumentValue) pending.argumentsText = argumentValue
    this.pendingCalls.delete(id)
    this.recordToolCall(pending)
  }

  private completeUsage(response: Readonly<Record<string, unknown>>): void {
    const usage = recordValue(response.usage)
    const inputDetails = recordValue(usage.input_tokens_details)
    const outputDetails = recordValue(usage.output_tokens_details)
    const cacheReadTokens = finiteNumber(usage.cache_read_tokens)
      ?? finiteNumber(inputDetails.cached_tokens)
    const cacheCreationTokens = finiteNumber(usage.cache_creation_tokens)
      ?? finiteNumber(outputDetails.cache_creation_tokens)
    const reasoningTokens = finiteNumber(outputDetails.reasoning_tokens)
    this.completePendingCalls()
    this.usage = {
      inputTokens: finiteNumber(usage.input_tokens) ?? 0,
      outputTokens: finiteNumber(usage.output_tokens) ?? 0,
      toolCalls: this.usage.toolCalls,
      finishReason: stringValue(response.status) || 'stop',
      ...(cacheReadTokens === undefined ? {} : { cacheReadTokens }),
      ...(cacheCreationTokens === undefined ? {} : { cacheCreationTokens }),
      ...(reasoningTokens === undefined ? {} : { reasoningTokens }),
    }
  }

  private completePendingCalls(): void {
    for (const [id, pending] of this.pendingCalls) {
      this.pendingCalls.delete(id)
      this.recordToolCall(pending)
    }
  }

  private recordToolCall(pending: PendingFunctionCall): void {
    if (!pending.name) return
    const call: ToolCall = {
      id: pending.id,
      type: 'function',
      function: {
        name: pending.name,
        arguments: parseToolArguments(pending.argumentsText),
      },
    }
    this.usage = { ...this.usage, toolCalls: [...this.usage.toolCalls, call] }
  }

  private completionDelta(): LlmDelta {
    return {
      finishReason: this.usage.finishReason,
      usage: tokenUsage(this.usage),
      ...(this.usage.toolCalls.length ? { toolCalls: this.usage.toolCalls } : {}),
    }
  }

  private createPendingCall(id: string): PendingFunctionCall {
    return { id, name: '', argumentsText: '' }
  }
}

function isFunctionCallItem(item: Readonly<Record<string, unknown>>): boolean {
  const type = stringValue(item.type)
  return type === 'function_call' || type === 'tool_call'
}

function argumentsText(value: unknown): string {
  if (typeof value === 'string') return value
  if (isJsonObject(value)) return JSON.stringify(value)
  return ''
}

function tokenUsage(usage: ResponsesUsage): TokenUsage {
  return {
    inputTokens: usage.inputTokens,
    outputTokens: usage.outputTokens,
    ...(usage.cacheReadTokens === undefined ? {} : { cacheReadTokens: usage.cacheReadTokens }),
    ...(usage.cacheCreationTokens === undefined ? {} : { cacheCreationTokens: usage.cacheCreationTokens }),
    ...(usage.reasoningTokens === undefined ? {} : { reasoningTokens: usage.reasoningTokens }),
  }
}

function finiteNumber(value: unknown): number | undefined {
  return typeof value === 'number' && Number.isFinite(value) ? value : undefined
}

function recordValue(value: unknown): Readonly<Record<string, unknown>> {
  return typeof value === 'object' && value !== null && !Array.isArray(value)
    ? value as Record<string, unknown>
    : {}
}

function stringValue(value: unknown): string {
  return typeof value === 'string' ? value : ''
}
